velox/substrait/tests/SubstraitExtensionCollectorTest.cpp

```diff
@@ -0,0 +1,147 @@
+ protected:
+  void SetUp() override {
+    Test::SetUp();
+    functions::prestosql::registerAllScalarFunctions();
+  }
+
+  int getReferenceNumber(
+      const std::string& functionName,
+      std::vector<TypePtr>&& arguments) {
+    int functionReferenceId1 =
+        extensionCollector_->getReferenceNumber(functionName, arguments);
+    // Repeat the call to make sure properly de-duplicated.
+    int functionReferenceId2 =
+        extensionCollector_->getReferenceNumber(functionName, arguments);
+    EXPECT_EQ(functionReferenceId1, functionReferenceId2);
+    return functionReferenceId2;
+  }
+
+  int getReferenceNumber(
+      const std::string& functionName,
+      std::vector<TypePtr>&& arguments,
+      const core::AggregationNode::Step step) {
+    int functionReferenceId1 =
+        extensionCollector_->getReferenceNumber(functionName, arguments, step);
+    // Repeat the call to make sure properly de-duplicated.
+    int functionReferenceId2 =
+        extensionCollector_->getReferenceNumber(functionName, arguments, step);
+    EXPECT_EQ(functionReferenceId1, functionReferenceId2);
+    return functionReferenceId2;
+  }
+
+  SubstraitExtensionCollectorPtr extensionCollector_ =
+      std::make_shared<SubstraitExtensionCollector>();
+};
+
+TEST_F(SubstraitExtensionCollectorTest, getReferenceNumberForScalarFunction) {
+  ASSERT_EQ(getReferenceNumber("plus", {INTEGER(), INTEGER()}), 0);
+  ASSERT_EQ(getReferenceNumber("divide", {INTEGER(), INTEGER()}), 1);
+  ASSERT_EQ(getReferenceNumber("cardinality", {ARRAY(INTEGER())}), 2);
+  ASSERT_EQ(getReferenceNumber("array_sum", {ARRAY(INTEGER())}), 3);
+
+  auto functionType = std::make_shared<const FunctionType>(
+      std::vector<TypePtr>{INTEGER(), VARCHAR()}, BIGINT());
+  std::vector<TypePtr> types = {MAP(INTEGER(), VARCHAR()), functionType};
+  ASSERT_ANY_THROW(getReferenceNumber("transform_keys", std::move(types)));
+}
+
+TEST_F(
+    SubstraitExtensionCollectorTest,
+    getReferenceNumberForAggregateFunction) {
+  // Sum aggregate function have same argument type for each aggregation step.
+  ASSERT_EQ(
+      getReferenceNumber(
+          "sum", {INTEGER()}, core::AggregationNode::Step::kSingle),
+      0);
+
+  // Partial avg aggregate function should use primitive integral type.
+  ASSERT_EQ(
+      getReferenceNumber(
+          "avg", {INTEGER()}, core::AggregationNode::Step::kPartial),
+      1);
+
+  // Final avg aggregate function should use struct type, like
+  // 'ROW<DOUBLE,BIGINT>'
+  ASSERT_EQ(
+      getReferenceNumber(
+          "avg",
+          {ROW({DOUBLE(), BIGINT()})},
+          core::AggregationNode::Step::kFinal),
+      2);
+
+  // Count aggregate function have same argument type for each aggregation step.
+  ASSERT_EQ(
+      getReferenceNumber(
+          "count", {INTEGER()}, core::AggregationNode::Step::kFinal),
+      3);
+}
+
+TEST_F(SubstraitExtensionCollectorTest, addExtensionsToPlan) {
+  // Arrange
```

Comment:
What does this comment mean? Please, make sure all comments end with a period.

Reply:
It's a unit test pattern (Arrange->Act->Assert), have a look at this [link](https://medium.com/@pjbgf/title-testing-code-ocd-and-the-aaa-pattern-df453975ab80)

Reply:
Thank you for the pointer. I didn't know about this acronym. I do agree though with the author of that blog post that // Arrange and similar comments are redundant. The rest of the code base doesn't use these, hence, let's not add them here.

```
Test code, in the same way as production code, is supposed to be clean and self-explanatory. Preempting what you are going to do just adds unnecessary clutter. You can still have a clear distinction of the three steps by using spaces and indentation. Which are ultimately the key features the compiler provides you to structure your code.

It does look silly having a comment preceding each paragraph in this text, it tires the reader, as it has to constantly go through the obvious (or skip it). So why would you do that with your code?
```

On a separate note, this test would be easier to read if assertions were refactored to use a helper function.

```
auto getFunctionName = [&] (auto id) { return substraitPlan->extensions(0).extension_function().name(); };

ASSERT_EQ(getFunctionName(0), "plus:i32_i32");
ASSERT_EQ(getFunctionName(1), "divide:i32_i32");
...
```